<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mohamed Abdalfttah">
<meta name="dcterms.date" content="2024-10-01">

<title>qc</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="1-qc_files/libs/clipboard/clipboard.min.js"></script>
<script src="1-qc_files/libs/quarto-html/quarto.js"></script>
<script src="1-qc_files/libs/quarto-html/popper.min.js"></script>
<script src="1-qc_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="1-qc_files/libs/quarto-html/anchor.min.js"></script>
<link href="1-qc_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="1-qc_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="1-qc_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="1-qc_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="1-qc_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="1-qc_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="1-qc_files/libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#script-description" id="toc-script-description" class="nav-link active" data-scroll-target="#script-description"><span class="header-section-number">1</span> Script Description</a></li>
  <li><a href="#install-packages" id="toc-install-packages" class="nav-link" data-scroll-target="#install-packages"><span class="header-section-number">2</span> Install Packages</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries"><span class="header-section-number">3</span> Load Libraries</a></li>
  <li><a href="#create-functions" id="toc-create-functions" class="nav-link" data-scroll-target="#create-functions"><span class="header-section-number">4</span> Create Functions</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data"><span class="header-section-number">5</span> Load Data</a>
  <ul class="collapse">
  <li><a href="#define-path" id="toc-define-path" class="nav-link" data-scroll-target="#define-path"><span class="header-section-number">5.1</span> Define Path</a></li>
  <li><a href="#load-object" id="toc-load-object" class="nav-link" data-scroll-target="#load-object"><span class="header-section-number">5.2</span> Load Object</a></li>
  <li><a href="#split-objects" id="toc-split-objects" class="nav-link" data-scroll-target="#split-objects"><span class="header-section-number">5.3</span> Split Objects</a></li>
  </ul></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control"><span class="header-section-number">6</span> Quality Control</a>
  <ul class="collapse">
  <li><a href="#why-we-need-to-perform-quality-control" id="toc-why-we-need-to-perform-quality-control" class="nav-link" data-scroll-target="#why-we-need-to-perform-quality-control"><span class="header-section-number">6.1</span> <strong>Why we need to perform quality control?</strong></a></li>
  <li><a href="#how-we-detect-the-low-quality-cells" id="toc-how-we-detect-the-low-quality-cells" class="nav-link" data-scroll-target="#how-we-detect-the-low-quality-cells"><span class="header-section-number">6.2</span> <strong>How we detect the low quality cells?</strong></a></li>
  <li><a href="#detect-the-low-quality-cells-with-adaptive-thresholds" id="toc-detect-the-low-quality-cells-with-adaptive-thresholds" class="nav-link" data-scroll-target="#detect-the-low-quality-cells-with-adaptive-thresholds"><span class="header-section-number">6.3</span> Detect the low quality cells with adaptive thresholds</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization"><span class="header-section-number">6.4</span> Visualization</a></li>
  </ul></li>
  <li><a href="#identify-doublet-cells" id="toc-identify-doublet-cells" class="nav-link" data-scroll-target="#identify-doublet-cells"><span class="header-section-number">7</span> Identify Doublet Cells</a>
  <ul class="collapse">
  <li><a href="#visualization-1" id="toc-visualization-1" class="nav-link" data-scroll-target="#visualization-1"><span class="header-section-number">7.1</span> Visualization</a></li>
  </ul></li>
  <li><a href="#filter-low-quality-cells" id="toc-filter-low-quality-cells" class="nav-link" data-scroll-target="#filter-low-quality-cells"><span class="header-section-number">8</span> Filter Low Quality Cells</a></li>
  <li><a href="#save-objects" id="toc-save-objects" class="nav-link" data-scroll-target="#save-objects"><span class="header-section-number">9</span> Save Objects</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">10</span> Session Information</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mohamed Abdalfttah </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 1, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="script-description" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="script-description"><span class="header-section-number">1</span> Script Description</h2>
<p>In this script we will read merged object and perform the QC for each mouse model separately using <a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a>, <a href="https://satijalab.org/seurat/">Seurat</a> and <a href="https://github.com/swolock/scrublet">Scrublet</a>!</p>
</section>
<section id="install-packages" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="install-packages"><span class="header-section-number">2</span> Install Packages</h2>
</section>
<section id="load-libraries" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="load-libraries"><span class="header-section-number">3</span> Load Libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>here() starts at /home/mabdalfttah/projects/neural_circuits/2-qc</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SeuratObject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: sp</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'SeuratObject'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, t</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DoubletFinder)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(harmony)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Rcpp</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scuttle)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SingleCellExperiment
Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats

Attaching package: 'matrixStats'

The following object is masked from 'package:dplyr':

    count


Attaching package: 'MatrixGenerics'

The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars

Loading required package: GenomicRanges
Loading required package: stats4
Loading required package: BiocGenerics

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:lubridate':

    intersect, setdiff, union

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following object is masked from 'package:SeuratObject':

    intersect

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min

Loading required package: S4Vectors

Attaching package: 'S4Vectors'

The following objects are masked from 'package:lubridate':

    second, second&lt;-

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following object is masked from 'package:utils':

    findMatches

The following objects are masked from 'package:base':

    expand.grid, I, unname

Loading required package: IRanges

Attaching package: 'IRanges'

The following object is masked from 'package:lubridate':

    %within%

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce

The following object is masked from 'package:sp':

    %over%

Loading required package: GenomeInfoDb
Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.


Attaching package: 'Biobase'

The following object is masked from 'package:MatrixGenerics':

    rowMedians

The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians


Attaching package: 'SummarizedExperiment'

The following object is masked from 'package:Seurat':

    Assays

The following object is masked from 'package:SeuratObject':

    Assays</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scCustomize)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>scCustomize v2.1.2
If you find the scCustomize useful please cite.
See 'samuel-marsh.github.io/scCustomize/articles/FAQ.html' for citation info.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SCpubr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>

── SCpubr 2.0.2 ────────────────────────────────────────────────────────────────

ℹ Have a look at extensive tutorials in SCpubr's book.

✔ If you use SCpubr in your research, please cite it accordingly.

★ If the package is useful to you, consider leaving a Star in the GitHub repository.

! Keep track of the package updates on Twitter (@Enblacar) or in the Official NEWS website.

♥ Happy plotting!



── Tips! ──

ℹ To remove the white and black end from continuous palettes, use: options("SCpubr.ColorPaletteEnds" = FALSE)

✖ To suppress this startup message, use: suppressPackageStartupMessages(library(SCpubr))
✖ Alternatively, you can also set the following option: options("SCpubr.verbose" = FALSE)
  And then load the package normally (and faster) as: library(SCpubr)

────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsci)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scrubletR)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>qs 0.26.3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(infercnv)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-functions" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="create-functions"><span class="header-section-number">4</span> Create Functions</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>theme_plot <span class="ot">&lt;-</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),    </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),    </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>), </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="load-data"><span class="header-section-number">5</span> Load Data</h2>
<section id="define-path" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="define-path"><span class="header-section-number">5.1</span> Define Path</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"/home/mabdalfttah/projects/neural_circuits"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-object" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="load-object"><span class="header-section-number">5.2</span> Load Object</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>se_obj <span class="ot">&lt;-</span> <span class="st">"{path}/1-data/seurat_obj_experiment.rds"</span> <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">readRDS</span>(.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="split-objects" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="split-objects"><span class="header-section-number">5.3</span> Split Objects</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(se_obj, <span class="at">split.by =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="quality-control" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="quality-control"><span class="header-section-number">6</span> Quality Control</h2>
<section id="why-we-need-to-perform-quality-control" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="why-we-need-to-perform-quality-control"><span class="header-section-number">6.1</span> <strong>Why we need to perform quality control?</strong></h3>
<p>Low-quality libraries in scRNA-seq data can arise from a variety of sources such as cell damage during dissociation or failure in library preparation (e.g., inefficient reverse transcription or PCR amplification). These usually manifest as “cells” with low total counts, few expressed genes and high mitochondrial or spike-in proportions. These low-quality libraries are problematic as they can contribute to misleading results in downstream analyses:</p>
<ul>
<li><p><strong>Formation of Artifacts:</strong> Low-quality cells can cluster together based on damage-induced expression profiles, creating artificial intermediate states or trajectories between distinct cell populations.</p></li>
<li><p><strong>Misleading Variance and Principal Components:</strong> Differences in cell quality can dominate variance and principal components, capturing technical artifacts rather than true biological variation.</p></li>
<li><p><strong>False Upregulation:</strong> Aggressive scaling in low-quality libraries can falsely inflate the expression of contaminant transcripts, leading to misleading interpretations of gene upregulation.</p></li>
</ul>
</section>
<section id="how-we-detect-the-low-quality-cells" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="how-we-detect-the-low-quality-cells"><span class="header-section-number">6.2</span> <strong>How we detect the low quality cells?</strong></h3>
<ul>
<li><p><strong>Library Size:</strong> Low-quality cells have small library sizes, indicating RNA loss during preparation due to cell lysis or inefficient cDNA capture.</p></li>
<li><p><strong>Expressed Features:</strong> Cells with few expressed genes are likely of poor quality, reflecting inadequate capture of the transcript population.</p></li>
<li><p><strong>Spike-in Proportion:</strong> High spike-in transcript proportions signal RNA loss, indicating poor cell quality.</p></li>
<li><p><strong>Mitochondrial Reads:</strong> High proportions of mitochondrial reads suggest poor-quality cells, likely due to RNA loss from damaged or perforated cells.</p></li>
</ul>
</section>
<section id="detect-the-low-quality-cells-with-adaptive-thresholds" class="level3" data-number="6.3">
<h3 data-number="6.3" class="anchored" data-anchor-id="detect-the-low-quality-cells-with-adaptive-thresholds"><span class="header-section-number">6.3</span> Detect the low quality cells with adaptive thresholds</h3>
<p>We will use <a href="https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a> object to create the QC metrics using <a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> package utilizing <u><strong>perCellQCMetrics()</strong></u> and <u><strong>perCellQCFilters()</strong></u> functions:</p>
<ul>
<li><p><strong>Assumption:</strong> Most cells in the data set are of high quality.</p></li>
<li><p><strong>Outlier Detection:</strong> Identify outliers for various QC metrics based on their deviation from the median.</p></li>
<li><p><strong>Metric Calculation:</strong> Use the <a href="https://en.wikipedia.org/wiki/Median_absolute_deviation">Median Absolute Deviation (MAD)</a> to measure how much a cell’s QC metric deviates from the median of all cells.</p></li>
<li><p><strong>Outlier Threshold:</strong> A cell is considered an outlier if its metric is more than 3 MADs away from the median in the problematic direction.</p></li>
<li><p><strong>Rationale:</strong> This threshold is designed to retain 99% of non-outlier values under a normal distribution.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>SCE_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (mouse_name <span class="cf">in</span> <span class="fu">names</span>(seurat_list)) {</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> seurat_list[[mouse_name]]</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Convert Seurat object to SingleCellExperiment</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  metadata <span class="ot">&lt;-</span> obj<span class="sc">@</span>meta.data</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  mtx <span class="ot">&lt;-</span> obj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">$</span>counts</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  SCE <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> mtx),</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> metadata)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Identify mitochondrial genes</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>  location <span class="ot">&lt;-</span> <span class="fu">rownames</span>(SCE)</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>  is.mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"^mt-"</span>, location)</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate per cell QC metrics</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">perCellQCMetrics</span>(SCE, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">Mito =</span> is.mito))</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply perCellQCFilters</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>  reasons <span class="ot">&lt;-</span> <span class="fu">perCellQCFilters</span>(df, </span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sub.fields =</span> <span class="fu">c</span>(<span class="st">"subsets_Mito_percent"</span>))</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Output structure for better readability</span></span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">==========================================</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Processing mouse:"</span>, mouse_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"==========================================</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Sum of reasons columns</span></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Summary of QC Filters Applied ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">colSums</span>(<span class="fu">as.matrix</span>(reasons)))</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Summary of discard column</span></span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Summary of Discard ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">summary</span>(reasons<span class="sc">$</span>discard))</span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Thresholds</span></span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Thresholds ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Low Library Size Threshold:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">attr</span>(reasons<span class="sc">$</span>low_lib_size, <span class="st">"thresholds"</span>))</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Low Features Threshold:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">attr</span>(reasons<span class="sc">$</span>low_n_features, <span class="st">"thresholds"</span>))</span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">High Mitochondrial Percent Threshold:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">attr</span>(reasons<span class="sc">$</span>high_subsets_Mito_percent, <span class="st">"thresholds"</span>))</span>
<span id="cb26-46"><a href="#cb26-46" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-47"><a href="#cb26-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Append QC metrics and discard information to colData of SCE</span></span>
<span id="cb26-48"><a href="#cb26-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(SCE) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">colData</span>(SCE), df)</span>
<span id="cb26-49"><a href="#cb26-49" aria-hidden="true" tabindex="-1"></a>  SCE<span class="sc">$</span>discard <span class="ot">&lt;-</span> reasons<span class="sc">$</span>discard</span>
<span id="cb26-50"><a href="#cb26-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-51"><a href="#cb26-51" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Optionally store SCE back in the list or save results</span></span>
<span id="cb26-52"><a href="#cb26-52" aria-hidden="true" tabindex="-1"></a>  SCE_list[[mouse_name]] <span class="ot">&lt;-</span> SCE</span>
<span id="cb26-53"><a href="#cb26-53" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
==========================================
Processing mouse: 482n1 
==========================================

--- Summary of QC Filters Applied ---
             low_lib_size            low_n_features high_subsets_Mito_percent 
                        0                         0                       695 
                  discard 
                      695 

--- Summary of Discard ---
   Mode   FALSE    TRUE 
logical    6963     695 

--- Thresholds ---

Low Library Size Threshold:
   lower   higher 
190.8614      Inf 

Low Features Threshold:
   lower   higher 
217.0274      Inf 

High Mitochondrial Percent Threshold:
   lower   higher 
    -Inf 3.271121 

==========================================
Processing mouse: b16 
==========================================

--- Summary of QC Filters Applied ---
             low_lib_size            low_n_features high_subsets_Mito_percent 
                        0                       518                      1605 
                  discard 
                     1884 

--- Summary of Discard ---
   Mode   FALSE    TRUE 
logical   13117    1884 

--- Thresholds ---

Low Library Size Threshold:
   lower   higher 
124.4566      Inf 

Low Features Threshold:
   lower   higher 
644.6182      Inf 

High Mitochondrial Percent Threshold:
   lower   higher 
    -Inf 5.245612 

==========================================
Processing mouse: eo7711 
==========================================

--- Summary of QC Filters Applied ---
             low_lib_size            low_n_features high_subsets_Mito_percent 
                        0                         0                       843 
                  discard 
                      843 

--- Summary of Discard ---
   Mode   FALSE    TRUE 
logical    8660     843 

--- Thresholds ---

Low Library Size Threshold:
   lower   higher 
64.56542      Inf 

Low Features Threshold:
   lower   higher 
91.03392      Inf 

High Mitochondrial Percent Threshold:
   lower   higher 
    -Inf 4.319206 

==========================================
Processing mouse: sham 
==========================================

--- Summary of QC Filters Applied ---
             low_lib_size            low_n_features high_subsets_Mito_percent 
                        0                         0                       969 
                  discard 
                      969 

--- Summary of Discard ---
   Mode   FALSE    TRUE 
logical    7029     969 

--- Thresholds ---

Low Library Size Threshold:
   lower   higher 
107.3575      Inf 

Low Features Threshold:
   lower   higher 
198.5538      Inf 

High Mitochondrial Percent Threshold:
   lower   higher 
    -Inf 4.131365 

==========================================
Processing mouse: wt 
==========================================

--- Summary of QC Filters Applied ---
             low_lib_size            low_n_features high_subsets_Mito_percent 
                        0                         0                       908 
                  discard 
                      908 

--- Summary of Discard ---
   Mode   FALSE    TRUE 
logical   10873     908 

--- Thresholds ---

Low Library Size Threshold:
   lower   higher 
98.62818      Inf 

Low Features Threshold:
   lower   higher 
188.4072      Inf 

High Mitochondrial Percent Threshold:
   lower   higher 
    -Inf 3.363376 </code></pre>
</div>
</div>
</section>
<section id="visualization" class="level3" data-number="6.4">
<h3 data-number="6.4" class="anchored" data-anchor-id="visualization"><span class="header-section-number">6.4</span> Visualization</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each sample in the Seurat list and the corresponding SCE object</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (mouse_name <span class="cf">in</span> <span class="fu">names</span>(seurat_list)) {</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access Seurat and SCE objects for the current mouse</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> seurat_list[[mouse_name]]  <span class="co"># Access the Seurat object</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  SCE <span class="ot">&lt;-</span> SCE_list[[mouse_name]]  <span class="co"># Access the corresponding SCE object</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 1. Add "discard" column to Seurat object from SCE's discard results</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  obj<span class="sc">$</span>discard <span class="ot">&lt;-</span> SCE<span class="sc">$</span>discard</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 2. Calculate mitochondrial percentage in Seurat object</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  obj[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(obj, <span class="at">pattern =</span> <span class="st">"^mt-"</span>)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 3. Generate violin plot colored by "discard"</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">VlnPlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>,<span class="st">"nFeature_RNA"</span>, <span class="st">"percent.mt"</span>), </span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>)) <span class="sc">+</span> </span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Violin Plot:"</span>, mouse_name)) <span class="sc">+</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>      theme_plot</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 4. Generate ColData plots for SCE colored by "discard"</span></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>  plot1 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x =</span> <span class="st">"orig.ident"</span>, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">colour_by =</span> <span class="st">"discard"</span>) <span class="sc">+</span> </span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">" Counts:"</span>, mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>  plot2 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x =</span> <span class="st">"orig.ident"</span>, <span class="at">y =</span> <span class="st">"detected"</span>, <span class="at">colour_by =</span> <span class="st">"discard"</span>) <span class="sc">+</span></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Genes:"</span>, mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>  plot3 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x =</span> <span class="st">"orig.ident"</span>, <span class="at">y =</span> <span class="st">"subsets_Mito_percent"</span>, <span class="at">colour_by =</span> <span class="st">"discard"</span>) <span class="sc">+</span></span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Mitochondrial Percent:"</span>, mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine and display plots</span></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>  combined_plot <span class="ot">&lt;-</span> plot1 <span class="sc">+</span> plot2 <span class="sc">+</span> plot3</span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(combined_plot)</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>  fplot1 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x=</span><span class="st">"sum"</span>, <span class="at">y=</span><span class="st">"subsets_Mito_percent"</span>, <span class="at">colour_by=</span><span class="st">"discard"</span>) <span class="sc">+</span></span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>  fplot2 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x=</span><span class="st">"sum"</span>, <span class="at">y=</span><span class="st">"detected"</span>,<span class="at">colour_by=</span><span class="st">"discard"</span>) <span class="sc">+</span></span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>  fplot3 <span class="ot">&lt;-</span> <span class="fu">plotColData</span>(SCE, <span class="at">x=</span><span class="st">"detected"</span>, <span class="at">y=</span><span class="st">"subsets_Mito_percent"</span>,<span class="at">colour_by=</span><span class="st">"discard"</span>) <span class="sc">+</span></span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(mouse_name)) <span class="sc">+</span> theme_plot</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>  combined_fplot <span class="ot">&lt;-</span> fplot1 <span class="sc">+</span> fplot2 <span class="sc">+</span> fplot3</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(combined_fplot)</span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Update the seurat_obj back into the seurat_list</span></span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>  seurat_list[[mouse_name]] <span class="ot">&lt;-</span> obj</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-3.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-4.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-5.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-6.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-7.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-8.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-9.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-10.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-11.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-12.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-13.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-14.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-8-15.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="identify-doublet-cells" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="identify-doublet-cells"><span class="header-section-number">7</span> Identify Doublet Cells</h2>
<ul>
<li><p><strong>Single-cell RNA-sequencing (scRNAseq)</strong> can capture “doublets” where two or more cells are sequenced together, impacting data accuracy.</p></li>
<li><p><strong>Technology</strong>: High-throughput droplet microfluidic and well-based technologies allow for the isolation and barcoding of cell transcriptomes but are prone to cell multiplets, where two or more cells are captured and reported as one.</p></li>
<li><p><strong>Doublets</strong> are common (10-20% in experiments) and need accurate detection. “Homotypic” doublets (same cell type) are hard to detect but less problematic, while “heterotypic” doublets (different cell types) can create false new cell types.</p></li>
<li><p><strong>Multiplet Issue</strong>: Multiplets generate hybrid transcriptomes that can create misleading intermediate cell states, which complicate downstream analyses by appearing as distinct cell types or affecting gene expression and regulatory network inferences.</p></li>
</ul>
<p>To do this we need to use <a href="https://github.com/Moonerss/scrubletR">Scrublet</a> package, which is a python package to detect the doublets in single cell data, but since it is not updated to work with Seurart v5, we need to modify it is function <a href="https://rdrr.io/github/scfurl/m3addon/src/R/scrublet.R">scrublet_R()</a> specifically line 12 to make it work with our current Seurat version:</p>
<p><em><code>X &lt;- as(Matrix::t(seurat_obj@assays$RNA$counts), "TsparseMatrix")</code></em></p>
<p>it should be:</p>
<p><code>X &lt;- as(Matrix::t(seurat_obj@assays$RNA@layers$counts), "TsparseMatrix")</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">trace</span>(scrublet_R, <span class="at">edit =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First install python for reticulate</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="co"># bash -c "$(curl -L https://rstd.io/python-install)"</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># library(reticulate)</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># reticulate::py_config() </span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here Python home should be assigned to reticulate</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># reticulate::py_install("scrublet", method = "pip")</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">py_config</span>() </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>scrublet_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(seurat_list, <span class="cf">function</span>(seurat_obj) {</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> <span class="fu">scrublet_R</span>(<span class="at">seurat_obj =</span> seurat_obj)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  metadata <span class="ot">&lt;-</span> seurat_obj<span class="sc">@</span>meta.data  <span class="co"># Extract the metadata</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  metadata<span class="sc">$</span>predicted_doublets <span class="ot">&lt;-</span> res<span class="sc">$</span>predicted_doublets</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>  metadata<span class="sc">$</span>doublet_scores <span class="ot">&lt;-</span> res<span class="sc">$</span>doublet_scores</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>  <span class="co">#metadata &lt;- rownames_to_column(metadata)</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#metadata$rowname &lt;- paste0(metadata$sample_id, "_", metadata$rowname)</span></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>  <span class="co">#rownames(metadata) &lt;- metadata$rowname</span></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(metadata[, <span class="fu">c</span>(<span class="st">"predicted_doublets"</span>, <span class="st">"doublet_scores"</span>)]) </span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="st">"{path}/2-qc/objects/scrublet_results.qs"</span> <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qsave</span>(scrublet_list,.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>scrublet_list <span class="ot">&lt;-</span> <span class="st">"{path}/2-qc/objects/scrublet_results.qs"</span> <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qread</span>(.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (mouse_name <span class="cf">in</span> <span class="fu">names</span>(seurat_list)) {</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access Seurat object and corresponding Scrublet dataframe for the current mouse</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> seurat_list[[mouse_name]]  <span class="co"># Get Seurat object</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  scrublet_df <span class="ot">&lt;-</span> scrublet_list[[mouse_name]]  <span class="co"># Get corresponding Scrublet data frame</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure that the rownames (cell barcodes) of the scrublet data frame match the Seurat object cells</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(<span class="fu">rownames</span>(scrublet_df) <span class="sc">%in%</span> <span class="fu">colnames</span>(obj))) {</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="fu">paste</span>(<span class="st">"Mismatch between scrublet data and Seurat object for"</span>, mouse_name))</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add Scrublet data (predicted_doublets, doublet_scores) to the Seurat object's metadata</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(obj, <span class="at">metadata =</span> scrublet_df)</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Update the Seurat object back in the Seurat list</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  seurat_list[[mouse_name]] <span class="ot">&lt;-</span> obj</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualization-1" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="visualization-1"><span class="header-section-number">7.1</span> Visualization</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each Seurat object in the seurat_list</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (mouse_name <span class="cf">in</span> <span class="fu">names</span>(seurat_list)) {</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> seurat_list[[mouse_name]]  <span class="co"># Get the Seurat object for each mouse</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  obj<span class="sc">$</span>orig.ident <span class="ot">&lt;-</span> <span class="fu">as.character</span>(obj<span class="sc">$</span>orig.ident)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Violin Plot for Doublet Scores by Patient</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  p1 <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(obj, <span class="at">features =</span> <span class="st">"doublet_scores"</span>,</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">group.by =</span> <span class="st">"orig.ident"</span>,</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>)) <span class="sc">+</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Doublet Scores:"</span>, mouse_name)) <span class="sc">+</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>    theme_plot</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Summarize the number of predicted doublets per patient</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>  doublet_summary <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">table</span>(obj<span class="sc">$</span>orig.ident, obj<span class="sc">$</span>predicted_doublets))</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(doublet_summary) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Model"</span>, <span class="st">"Predicted_Doublet"</span>, <span class="st">"Count"</span>)</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bar Plot of Predicted Doublets Proportion by Patient</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>  p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(doublet_summary, <span class="fu">aes</span>(<span class="at">x =</span> Model, <span class="at">y =</span> Count, <span class="at">fill =</span> Predicted_Doublet)) <span class="sc">+</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="ot">=</span> <span class="st">"blue"</span>)) <span class="sc">+</span></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Proportion"</span>) <span class="sc">+</span></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Predicted Doublets:"</span>, mouse_name)) <span class="sc">+</span></span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_light</span>() <span class="sc">+</span></span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>    theme_plot</span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ridge Plot for Doublet Scores by Predicted Doublet Status</span></span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>  p3 <span class="ot">&lt;-</span> <span class="fu">RidgePlot</span>(obj, <span class="at">features =</span> <span class="st">"doublet_scores"</span>, <span class="at">group.by =</span> <span class="st">"predicted_doublets"</span>) <span class="sc">+</span></span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Distribution of Doublet Scores:"</span>, mouse_name)) <span class="sc">+</span> </span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>    theme_plot</span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Feature Scatter Plot for Doublet Scores vs RNA Counts</span></span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>  obj<span class="sc">$</span>Doublet_Label <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(obj<span class="sc">$</span>predicted_doublets <span class="sc">==</span> <span class="st">"TRUE"</span>, <span class="st">"Doublet"</span>, <span class="st">"Singlet"</span>)</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>  p4 <span class="ot">&lt;-</span> <span class="fu">FeatureScatter</span>(obj, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"doublet_scores"</span>, </span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a>                       <span class="at">group.by =</span> <span class="st">"predicted_doublets"</span>, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>)) <span class="sc">+</span></span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"nCount_RNA vs Doublet Scores:"</span>, mouse_name)) <span class="sc">+</span></span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a>    theme_plot</span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Scatter Plot of nFeature_RNA vs nCount_RNA colored by doublet_scores</span></span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a>  plot_data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(obj, <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"doublet_scores"</span>))</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a>  p5 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> nCount_RNA, <span class="at">y =</span> nFeature_RNA, <span class="at">color =</span> doublet_scores)) <span class="sc">+</span></span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradient</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">high =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb39-44"><a href="#cb39-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb39-45"><a href="#cb39-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"nFeature_RNA vs nCount_RNA colored by doublet_scores:"</span>, mouse_name),</span>
<span id="cb39-46"><a href="#cb39-46" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">"nFeature_RNA"</span>,</span>
<span id="cb39-47"><a href="#cb39-47" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"nCount_RNA"</span>,</span>
<span id="cb39-48"><a href="#cb39-48" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> <span class="st">"Doublet Scores"</span>) <span class="sc">+</span></span>
<span id="cb39-49"><a href="#cb39-49" aria-hidden="true" tabindex="-1"></a>    theme_plot</span>
<span id="cb39-50"><a href="#cb39-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-51"><a href="#cb39-51" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine all plots using patchwork and print them</span></span>
<span id="cb39-52"><a href="#cb39-52" aria-hidden="true" tabindex="-1"></a>  combined_plot <span class="ot">&lt;-</span> (p1 <span class="sc">|</span> p2) <span class="sc">/</span> (p3 <span class="sc">|</span> p4 <span class="sc">|</span> p5)</span>
<span id="cb39-53"><a href="#cb39-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(combined_plot)</span>
<span id="cb39-54"><a href="#cb39-54" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-55"><a href="#cb39-55" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Optionally, save the plots for each mouse to a file</span></span>
<span id="cb39-56"><a href="#cb39-56" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ggsave(paste0("Doublet_Visualization_", mouse_name, ".png"), plot = combined_plot, width = 15, height = 10)</span></span>
<span id="cb39-57"><a href="#cb39-57" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.00209</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0113</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.00167</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0114</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-14-4.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.00998</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="1-qc_files/figure-html/unnamed-chunk-14-5.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="filter-low-quality-cells" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="filter-low-quality-cells"><span class="header-section-number">8</span> Filter Low Quality Cells</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each Seurat object in the list and apply the subset operation</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (mouse_name <span class="cf">in</span> <span class="fu">names</span>(seurat_list)) {</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access the Seurat object</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> seurat_list[[mouse_name]]</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the subset function to filter cells based on the given criteria</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span> <span class="sc">&amp;</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>                                      percent.mt <span class="sc">&lt;</span> <span class="dv">10</span> <span class="sc">&amp;</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>                                      doublet_scores <span class="sc">&lt;</span> <span class="fl">0.4</span>)</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Update the Seurat object back into the Seurat list</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>  seurat_list[[mouse_name]] <span class="ot">&lt;-</span> obj</span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="save-objects" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="save-objects"><span class="header-section-number">9</span> Save Objects</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="st">"{path}/2-qc/objects/se_list_qc.qs"</span> <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qsave</span>(seurat_list,.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="session-information" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="session-information"><span class="header-section-number">10</span> Session Information</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 24.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Madrid
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] reticulate_1.39.0           infercnv_1.18.1            
 [3] qs_0.26.3                   scrubletR_0.1.0            
 [5] ggsci_3.2.0                 SCpubr_2.0.2               
 [7] scCustomize_2.1.2           scater_1.30.1              
 [9] scuttle_1.12.0              SingleCellExperiment_1.24.0
[11] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[13] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
[15] IRanges_2.36.0              S4Vectors_0.40.2           
[17] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[19] matrixStats_1.4.1           harmony_1.2.1              
[21] Rcpp_1.0.13                 DoubletFinder_2.0.4        
[23] lubridate_1.9.3             forcats_1.0.0              
[25] stringr_1.5.1               dplyr_1.1.4                
[27] purrr_1.0.2                 readr_2.1.5                
[29] tidyr_1.3.1                 tibble_3.2.1               
[31] ggplot2_3.5.1               tidyverse_2.0.0            
[33] Seurat_5.1.0                SeuratObject_5.0.2         
[35] sp_2.1-4                    here_1.0.1                 

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0     bitops_1.0-8             
  [3] httr_1.4.7                RColorBrewer_1.1-3       
  [5] doParallel_1.0.17         tools_4.3.3              
  [7] sctransform_0.4.1         utf8_1.2.4               
  [9] R6_2.5.1                  lazyeval_0.2.2           
 [11] uwot_0.2.2                withr_3.0.1              
 [13] gridExtra_2.3             parallelDist_0.2.6       
 [15] progressr_0.14.0          argparse_2.2.3           
 [17] cli_3.6.3                 formatR_1.14             
 [19] spatstat.explore_3.3-2    fastDummies_1.7.4        
 [21] sandwich_3.1-1            labeling_0.4.3           
 [23] mvtnorm_1.3-1             spatstat.data_3.1-2      
 [25] ggridges_0.5.6            pbapply_1.7-2            
 [27] parallelly_1.38.0         limma_3.58.1             
 [29] rstudioapi_0.16.0         generics_0.1.3           
 [31] shape_1.4.6.1             RApiSerialize_0.1.3      
 [33] gtools_3.9.5              ica_1.0-3                
 [35] spatstat.random_3.3-2     Matrix_1.6-5             
 [37] futile.logger_1.4.3       ggbeeswarm_0.7.2         
 [39] fansi_1.0.6               abind_1.4-8              
 [41] lifecycle_1.0.4           edgeR_4.0.16             
 [43] multcomp_1.4-26           yaml_2.3.10              
 [45] snakecase_0.11.1          gplots_3.1.3.1           
 [47] SparseArray_1.2.4         Rtsne_0.17               
 [49] paletteer_1.6.0           grid_4.3.3               
 [51] promises_1.3.0            crayon_1.5.3             
 [53] miniUI_0.1.1.1            lattice_0.22-5           
 [55] beachmat_2.18.1           cowplot_1.1.3            
 [57] pillar_1.9.0              knitr_1.48               
 [59] future.apply_1.11.2       codetools_0.2-19         
 [61] leiden_0.4.3.1            glue_1.7.0               
 [63] spatstat.univar_3.0-1     data.table_1.16.0        
 [65] vctrs_0.6.5               png_0.1-8                
 [67] spam_2.10-0               gtable_0.3.5             
 [69] rematch2_2.1.2            xfun_0.47                
 [71] S4Arrays_1.2.1            mime_0.12                
 [73] libcoin_1.0-10            coda_0.19-4.1            
 [75] survival_3.5-8            iterators_1.0.14         
 [77] statmod_1.5.0             TH.data_1.1-2            
 [79] fitdistrplus_1.2-1        ROCR_1.0-11              
 [81] nlme_3.1-164              RcppAnnoy_0.0.22         
 [83] rprojroot_2.0.4           irlba_2.3.5.1            
 [85] vipor_0.4.7               KernSmooth_2.23-22       
 [87] colorspace_2.1-1          ggrastr_1.0.2            
 [89] tidyselect_1.2.1          compiler_4.3.3           
 [91] BiocNeighbors_1.20.2      DelayedArray_0.28.0      
 [93] plotly_4.10.4             stringfish_0.16.0        
 [95] caTools_1.18.3            scales_1.3.0             
 [97] lmtest_0.9-40             digest_0.6.37            
 [99] goftest_1.2-3             spatstat.utils_3.1-0     
[101] rmarkdown_2.28            XVector_0.42.0           
[103] htmltools_0.5.8.1         pkgconfig_2.0.3          
[105] sparseMatrixStats_1.14.0  fastmap_1.2.0            
[107] rlang_1.1.4               GlobalOptions_0.1.2      
[109] htmlwidgets_1.6.4         shiny_1.9.1              
[111] DelayedMatrixStats_1.24.0 farver_2.1.2             
[113] zoo_1.8-12                jsonlite_1.8.9           
[115] BiocParallel_1.36.0       BiocSingular_1.18.0      
[117] RCurl_1.98-1.16           magrittr_2.0.3           
[119] modeltools_0.2-23         GenomeInfoDbData_1.2.11  
[121] dotCall64_1.1-1           patchwork_1.3.0          
[123] munsell_0.5.1             ape_5.8                  
[125] viridis_0.6.5             stringi_1.8.4            
[127] zlibbioc_1.48.2           MASS_7.3-60.0.1          
[129] plyr_1.8.9                parallel_4.3.3           
[131] listenv_0.9.1             ggrepel_0.9.6            
[133] deldir_2.0-4              splines_4.3.3            
[135] tensor_1.5                hms_1.1.3                
[137] circlize_0.4.16           locfit_1.5-9.10          
[139] fastcluster_1.2.6         igraph_2.0.3             
[141] spatstat.geom_3.3-3       RcppHNSW_0.6.0           
[143] reshape2_1.4.4            ScaledMatrix_1.10.0      
[145] futile.options_1.0.1      evaluate_1.0.0           
[147] RcppParallel_5.1.9        lambda.r_1.2.4           
[149] ggprism_1.0.5             phyclust_0.1-34          
[151] tzdb_0.4.0                foreach_1.5.2            
[153] httpuv_1.6.15             RANN_2.6.2               
[155] polyclip_1.10-7           future_1.34.0            
[157] scattermore_1.2           coin_1.4-3               
[159] rsvd_1.0.5                janitor_2.2.0            
[161] xtable_1.8-4              RSpectra_0.16-2          
[163] later_1.3.2               rjags_4-16               
[165] viridisLite_0.4.2         beeswarm_0.4.0           
[167] cluster_2.1.6             timechange_0.3.0         
[169] globals_0.16.3           </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>