<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mohamed Abdalfttah">
<meta name="dcterms.date" content="2024-10-03">

<title>lung_annotation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="3-lung_annotation_files/libs/clipboard/clipboard.min.js"></script>
<script src="3-lung_annotation_files/libs/quarto-html/quarto.js"></script>
<script src="3-lung_annotation_files/libs/quarto-html/popper.min.js"></script>
<script src="3-lung_annotation_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="3-lung_annotation_files/libs/quarto-html/anchor.min.js"></script>
<link href="3-lung_annotation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="3-lung_annotation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="3-lung_annotation_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="3-lung_annotation_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="3-lung_annotation_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="3-lung_annotation_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="3-lung_annotation_files/libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#script-description" id="toc-script-description" class="nav-link active" data-scroll-target="#script-description"><span class="header-section-number">0.1</span> Script Description</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries"><span class="header-section-number">0.2</span> Load Libraries</a></li>
  <li><a href="#create-function" id="toc-create-function" class="nav-link" data-scroll-target="#create-function"><span class="header-section-number">0.3</span> Create Function</a></li>
  <li><a href="#load-objects" id="toc-load-objects" class="nav-link" data-scroll-target="#load-objects"><span class="header-section-number">0.4</span> Load Objects</a>
  <ul class="collapse">
  <li><a href="#select-the-model" id="toc-select-the-model" class="nav-link" data-scroll-target="#select-the-model"><span class="header-section-number">0.4.1</span> Select the model</a></li>
  <li><a href="#check-the-doublets" id="toc-check-the-doublets" class="nav-link" data-scroll-target="#check-the-doublets"><span class="header-section-number">0.4.2</span> Check the Doublets</a></li>
  </ul></li>
  <li><a href="#explaring-the-markers" id="toc-explaring-the-markers" class="nav-link" data-scroll-target="#explaring-the-markers"><span class="header-section-number">1</span> Explaring the Markers</a>
  <ul class="collapse">
  <li><a href="#how-people-annotate-cells" id="toc-how-people-annotate-cells" class="nav-link" data-scroll-target="#how-people-annotate-cells"><span class="header-section-number">1.1</span> How people annotate cells?</a></li>
  </ul></li>
  <li><a href="#calculate-marker-genes" id="toc-calculate-marker-genes" class="nav-link" data-scroll-target="#calculate-marker-genes"><span class="header-section-number">2</span> Calculate Marker Genes</a>
  <ul class="collapse">
  <li><a href="#heatmap-visulization" id="toc-heatmap-visulization" class="nav-link" data-scroll-target="#heatmap-visulization"><span class="header-section-number">2.1</span> Heatmap Visulization</a>
  <ul class="collapse">
  <li><a href="#save-marker-genes" id="toc-save-marker-genes" class="nav-link" data-scroll-target="#save-marker-genes"><span class="header-section-number">2.1.1</span> Save Marker Genes</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#assign-the-annotation" id="toc-assign-the-annotation" class="nav-link" data-scroll-target="#assign-the-annotation"><span class="header-section-number">3</span> Assign The Annotation</a>
  <ul class="collapse">
  <li><a href="#confirm-the-annotation" id="toc-confirm-the-annotation" class="nav-link" data-scroll-target="#confirm-the-annotation"><span class="header-section-number">3.1</span> Confirm the Annotation</a>
  <ul class="collapse">
  <li><a href="#dot-plot" id="toc-dot-plot" class="nav-link" data-scroll-target="#dot-plot"><span class="header-section-number">3.1.1</span> Dot plot</a></li>
  <li><a href="#clusterd-dot-plot" id="toc-clusterd-dot-plot" class="nav-link" data-scroll-target="#clusterd-dot-plot"><span class="header-section-number">3.1.2</span> Clusterd Dot plot</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#save-object" id="toc-save-object" class="nav-link" data-scroll-target="#save-object"><span class="header-section-number">4</span> Save Object</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">5</span> Session Information</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mohamed Abdalfttah </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 3, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="script-description" class="level2" data-number="0.1">
<h2 data-number="0.1" class="anchored" data-anchor-id="script-description"><span class="header-section-number">0.1</span> Script Description</h2>
<p>In this script we will annotate the cells from each mouse model separately using two approaches:</p>
<p>1- Canonical Markers</p>
<p>2- Gene expression profile</p>
<pre><code> "482n1" Lung
 "b16" Melanoma
 "eo7711" Breast</code></pre>
</section>
<section id="load-libraries" class="level2" data-number="0.2">
<h2 data-number="0.2" class="anchored" data-anchor-id="load-libraries"><span class="header-section-number">0.2</span> Load Libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("devtools")</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github("immunogenomics/presto")</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages('assertthat')</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("ComplexHeatmap")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>here() starts at /home/mabdalfttah/projects/neural_circuits/4-annotation</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SeuratObject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: sp</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'SeuratObject'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, t</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DoubletFinder)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(harmony)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Rcpp</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scuttle)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SingleCellExperiment
Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats

Attaching package: 'matrixStats'

The following object is masked from 'package:dplyr':

    count


Attaching package: 'MatrixGenerics'

The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars

Loading required package: GenomicRanges
Loading required package: stats4
Loading required package: BiocGenerics

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:lubridate':

    intersect, setdiff, union

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following object is masked from 'package:SeuratObject':

    intersect

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min

Loading required package: S4Vectors

Attaching package: 'S4Vectors'

The following objects are masked from 'package:lubridate':

    second, second&lt;-

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following object is masked from 'package:utils':

    findMatches

The following objects are masked from 'package:base':

    expand.grid, I, unname

Loading required package: IRanges

Attaching package: 'IRanges'

The following object is masked from 'package:lubridate':

    %within%

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce

The following object is masked from 'package:sp':

    %over%

Loading required package: GenomeInfoDb
Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.


Attaching package: 'Biobase'

The following object is masked from 'package:MatrixGenerics':

    rowMedians

The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians


Attaching package: 'SummarizedExperiment'

The following object is masked from 'package:Seurat':

    Assays

The following object is masked from 'package:SeuratObject':

    Assays</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scCustomize)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>scCustomize v2.1.2
If you find the scCustomize useful please cite.
See 'samuel-marsh.github.io/scCustomize/articles/FAQ.html' for citation info.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SCpubr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>

── SCpubr 2.0.2 ────────────────────────────────────────────────────────────────

ℹ Have a look at extensive tutorials in SCpubr's book.

✔ If you use SCpubr in your research, please cite it accordingly.

★ If the package is useful to you, consider leaving a Star in the GitHub repository.

! Keep track of the package updates on Twitter (@Enblacar) or in the Official NEWS website.

♥ Happy plotting!



── Tips! ──

ℹ To remove the white and black end from continuous palettes, use: options("SCpubr.ColorPaletteEnds" = FALSE)

✖ To suppress this startup message, use: suppressPackageStartupMessages(library(SCpubr))
✖ Alternatively, you can also set the following option: options("SCpubr.verbose" = FALSE)
  And then load the package normally (and faster) as: library(SCpubr)

────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsci)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scrubletR)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>qs 0.26.3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(infercnv)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: viridisLite</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'glue'

The following object is masked from 'package:SummarizedExperiment':

    trim

The following object is masked from 'package:GenomicRanges':

    trim

The following object is masked from 'package:IRanges':

    trim</code></pre>
</div>
</div>
</section>
<section id="create-function" class="level2" data-number="0.3">
<h2 data-number="0.3" class="anchored" data-anchor-id="create-function"><span class="header-section-number">0.3</span> Create Function</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>Save_Markers <span class="ot">=</span> <span class="cf">function</span>(Markers, Path_write){</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  mgs_ls <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">sort</span>(<span class="fu">unique</span>(Markers<span class="sc">$</span>cluster)), <span class="cf">function</span>(clust) {</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    Markers <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">filter</span>(cluster <span class="sc">==</span> clust <span class="sc">&amp;</span> pct<span class="fl">.1</span> <span class="sc">&gt;</span> <span class="fl">0.25</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(avg_log2FC))</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(mgs_ls) <span class="ot">=</span> <span class="fu">glue</span>(<span class="st">"cluster-{sort(unique(Markers$cluster))}"</span>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  openxlsx<span class="sc">::</span><span class="fu">write.xlsx</span>(mgs_ls, <span class="at">file =</span> Path_write)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-objects" class="level2" data-number="0.4">
<h2 data-number="0.4" class="anchored" data-anchor-id="load-objects"><span class="header-section-number">0.4</span> Load Objects</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"/home/mabdalfttah/projects/neural_circuits/"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="st">"{path}/3-preprocess/objects/se_list_preprocess.qs"</span> <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qread</span>(.)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>seurat_list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`482n1`
An object of class Seurat 
16974 features across 7557 samples within 1 assay 
Active assay: RNA (16974 features, 722 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap

$b16
An object of class Seurat 
16974 features across 14493 samples within 1 assay 
Active assay: RNA (16974 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap

$eo7711
An object of class Seurat 
16974 features across 9350 samples within 1 assay 
Active assay: RNA (16974 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap

$sham
An object of class Seurat 
16974 features across 7849 samples within 1 assay 
Active assay: RNA (16974 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap

$wt
An object of class Seurat 
16974 features across 11625 samples within 1 assay 
Active assay: RNA (16974 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<section id="select-the-model" class="level3" data-number="0.4.1">
<h3 data-number="0.4.1" class="anchored" data-anchor-id="select-the-model"><span class="header-section-number">0.4.1</span> Select the model</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="st">"482n1"</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>se_obj <span class="ot">&lt;-</span> seurat_list[[model]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(seurat_list)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(se_obj, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>), <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">label =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(se_obj, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>), <span class="at">group.by =</span> <span class="st">"RNA_snn_res.0.2"</span>, <span class="at">label =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(se_obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"percent.mt"</span>), <span class="at">pt.size =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-8-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="check-the-doublets" class="level3" data-number="0.4.2">
<h3 data-number="0.4.2" class="anchored" data-anchor-id="check-the-doublets"><span class="header-section-number">0.4.2</span> Check the Doublets</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(<span class="at">object =</span> se_obj, <span class="at">features =</span> <span class="st">"doublet_scores"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> se_obj, <span class="at">group.by =</span> <span class="st">"predicted_doublets"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-9-2.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="explaring-the-markers" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Explaring the Markers</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming seurat_obj is your Seurat object for visualization</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>brain_cell_types_markers <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Neurons =</span> <span class="fu">c</span>(<span class="st">"Snap25"</span>, <span class="st">"Rbfox3"</span>, <span class="st">"Syt1"</span>, <span class="st">"Grin1"</span>, <span class="st">"Gad1"</span>, <span class="st">"Gad2"</span>, <span class="st">"Slc17a7"</span>, <span class="st">"Map2"</span>),</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Astrocytes =</span> <span class="fu">c</span>(<span class="st">"Gfap"</span>, <span class="st">"Aqp4"</span>, <span class="st">"Aldh1l1"</span>, <span class="st">"Slc1a3"</span>, <span class="st">"S100b"</span>),</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Oligodendrocytes =</span> <span class="fu">c</span>(<span class="st">"Mog"</span>, <span class="st">"Olig2"</span>, <span class="st">"Plp1"</span>, <span class="st">"Cnp"</span>, <span class="st">"Mbpa"</span>),</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">Microglia =</span> <span class="fu">c</span>(<span class="st">"C1qa"</span>, <span class="st">"Csf1r"</span>, <span class="st">"P2ry12"</span>, <span class="st">"Aif1"</span>, <span class="st">"Tmem119"</span>),</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">OPCs =</span> <span class="fu">c</span>(<span class="st">"Pdgfra"</span>, <span class="st">"Cspg4"</span>, <span class="st">"Sox10"</span>, <span class="st">"Olig1"</span>, <span class="st">"Olig2"</span>),</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Endothelial =</span> <span class="fu">c</span>(<span class="st">"Flt1"</span>, <span class="st">"Pecam1"</span>, <span class="st">"Cldn5"</span>, <span class="st">"Cdh5"</span>, <span class="st">"Slco1c1"</span>),</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">Ependymal =</span> <span class="fu">c</span>(<span class="st">"Foxj1"</span>, <span class="st">"Vim"</span>, <span class="st">"Ttr"</span>, <span class="st">"Dnah5"</span>, <span class="st">"Sox9"</span>),</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">vSMCs =</span> <span class="fu">c</span>(<span class="st">"Acta2"</span>, <span class="st">"Tagln"</span>, <span class="st">"Myh11"</span>, <span class="st">"Cnn1"</span>, <span class="st">"Lmod1"</span>),</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">Pericytes =</span> <span class="fu">c</span>(<span class="st">"Pdgfrb"</span>, <span class="st">"Rgs5"</span>, <span class="st">"Anpep"</span>, <span class="st">"Cspg4"</span>, <span class="st">"Kcnj8"</span>),</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">Monocytes =</span> <span class="fu">c</span>(<span class="st">"Cd14"</span>, <span class="st">"Psap"</span>, <span class="st">"S100a9"</span>, <span class="st">"Fcer1g"</span>),</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">Epithilial =</span> <span class="fu">c</span>(<span class="st">"Krt18"</span>, <span class="st">"Epcam"</span>, <span class="st">"Krt19"</span>, <span class="st">"Krt23"</span>),</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">T_Cells =</span> <span class="fu">c</span>(<span class="st">"Ptprc"</span>, <span class="st">"Cd3e"</span>, <span class="st">"Cd3d"</span>, <span class="st">"Cd8a"</span>, <span class="st">"Cd3g"</span>),</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">NK =</span> <span class="fu">c</span>(<span class="st">"Ncr1"</span>, <span class="st">"Klra8"</span>, <span class="st">"Klri2"</span>, <span class="st">"Klrb1c"</span>),</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">Mesenchymal  =</span> <span class="fu">c</span>(<span class="st">"Dcn"</span>, <span class="st">"Col3a1"</span>, <span class="st">"Col1a1"</span>)</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each cell type and visualize the markers using FeaturePlot</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (cell_type <span class="cf">in</span> <span class="fu">names</span>(brain_cell_types_markers)) {</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>  markers <span class="ot">&lt;-</span> brain_cell_types_markers[[cell_type]]</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Visualizing"</span>, cell_type, <span class="st">"markers"</span>))</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">FeaturePlot</span>(se_obj, <span class="at">features =</span> markers, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)))  <span class="co"># Center the title</span></span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Neurons markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Astrocytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Oligodendrocytes markers"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following requested variables were not found: Mbpa</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-3.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Microglia markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-4.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing OPCs markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-5.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Endothelial markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-6.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Ependymal markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-7.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing vSMCs markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-8.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Pericytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-9.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Monocytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-10.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Epithilial markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-11.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing T_Cells markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-12.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing NK markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-13.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Mesenchymal markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-14.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each cell type and visualize the markers using FeaturePlot</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (cell_type <span class="cf">in</span> <span class="fu">names</span>(brain_cell_types_markers)) {</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  markers <span class="ot">&lt;-</span> brain_cell_types_markers[[cell_type]]</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Visualizing"</span>, cell_type, <span class="st">"markers"</span>))</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">VlnPlot</span>(se_obj, <span class="at">features =</span> markers,</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>), <span class="at">pt.size =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)))  <span class="co"># Center the title</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Neurons markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-15.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Astrocytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-16.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Oligodendrocytes markers"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following requested variables were not found: Mbpa</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-17.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Microglia markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-18.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing OPCs markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-19.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Endothelial markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-20.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Ependymal markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-21.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing vSMCs markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-22.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Pericytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-23.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Monocytes markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-24.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Epithilial markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-25.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing T_Cells markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-26.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing NK markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-27.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Visualizing Mesenchymal markers"</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-10-28.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
<section id="how-people-annotate-cells" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="how-people-annotate-cells"><span class="header-section-number">1.1</span> How people annotate cells?</h2>
<p>Here is this paper they annotate the major cell types like this: Annotation of cell clusters based on the expression of canonical gene markers identified the following populations: excitatory neurons (Ex) that express <em>SYT1</em> and <em>SLC17A7</em>, inhibitory neurons (In) that express <em>SYT1</em> and <em>GAD1</em>, astrocytes (Ast1 and Ast2) that express <em>AQP4</em>, ependymal cells (Ep) that express <em>CFAP299</em>, oligodendrocyte progenitor cell (OPC) that express <em>VCAN</em>, oligodendrocytes (Oli) that express <em>MOBP</em>, epithelial cells (Epi) that express <em>HTR2C</em>, endothelial cells (End) that express <em>FLT1</em>, mesenchymal cells (Mes) that express <em>COL1A1</em>, pericytes (Per1 and Per2) that express <em>PDGFRB</em>, microglial cells (Mic) that express <em>APBB1IP</em>, and lymphocyte (LM) that express <em>CD96</em> and monocytes/macrophage (Mo/MP), expressing <em>CD163</em> (<a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-021-00933-8">reference</a>)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">str_to_title</span>(<span class="fu">c</span>(<span class="st">"SYT1"</span>, <span class="co"># Neurons</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"SLC17A7"</span>, <span class="co"># excitatory neurons</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"GAD1"</span>, <span class="co"># inhibitory neurons</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"AQP4"</span>, <span class="co"># Astrocytes</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"CFAP299"</span>, <span class="co"># ependymal </span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"VCAN"</span>, <span class="co"># oligodendrocyte progenitor </span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"MOBP"</span>, <span class="co"># oligodendrocytes</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"Epcam"</span>, <span class="co"># epithelial </span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"FLT1"</span>, <span class="co"># endothelial </span></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"COL1A1"</span>, <span class="co"># mesenchymal </span></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"PDGFRB"</span>, <span class="co"># pericytes</span></span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"APBB1IP"</span>,<span class="co"># microglial </span></span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"Ptprc"</span>, <span class="co"># lymphocyte</span></span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"CD163"</span>, <span class="co"># monocytes/macrophage</span></span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>                                          <span class="st">"Mki67"</span>, <span class="st">"Top2a"</span> <span class="co">#Prolifrative Cells</span></span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>                                              ))</span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(se_obj, <span class="at">features =</span> genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(se_obj, <span class="at">features =</span> genes, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>),</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"RNA_snn_res.0.2"</span>,<span class="at">pt.size =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="calculate-marker-genes" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Calculate Marker Genes</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(se_obj) <span class="ot">&lt;-</span> se_obj<span class="sc">$</span>RNA_snn_res.<span class="fl">0.2</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> se_obj, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 9</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 11</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 12</code></pre>
</div>
</div>
<section id="heatmap-visulization" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="heatmap-visulization"><span class="header-section-number">2.1</span> Heatmap Visulization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>markers <span class="sc">%&gt;%</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="ot">-&gt;</span> top10</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(se_obj, <span class="at">features =</span> top10<span class="sc">$</span>gene) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<section id="save-marker-genes" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="save-marker-genes"><span class="header-section-number">2.1.1</span> Save Marker Genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="st">"{path}/4-annotation/objects/{model}_markers_res_0.2.xlsx"</span> <span class="sc">%&gt;%</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Save_Markers</span>(markers, <span class="at">Path_write =</span> .)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="assign-the-annotation" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Assign The Annotation</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>se_obj<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> se_obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">annotation_lvl1 =</span> <span class="fu">case_when</span>(</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">0</span> <span class="sc">~</span> <span class="st">"Excitatory neurons"</span>,</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"Microglia"</span>,</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">2</span> <span class="sc">~</span> <span class="st">"Microglia"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">3</span> <span class="sc">~</span> <span class="st">"Myeloid"</span>,</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Oligodendrocytes"</span>,</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Astrocytes"</span>,</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">6</span> <span class="sc">~</span> <span class="st">"Inhibitory neurons"</span>,</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">7</span> <span class="sc">~</span> <span class="st">"T Cells"</span>,</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">8</span> <span class="sc">~</span> <span class="st">"Epithelial"</span>,</span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">9</span> <span class="sc">~</span> <span class="st">"OPCs"</span>,</span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">10</span> <span class="sc">~</span> <span class="st">"NK Cells"</span>,</span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">11</span> <span class="sc">~</span> <span class="st">"Endothelial"</span>,</span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a>    RNA_snn_res.<span class="fl">0.2</span> <span class="sc">==</span> <span class="dv">12</span> <span class="sc">~</span> <span class="st">"Mesenchymal"</span></span>
<span id="cb88-16"><a href="#cb88-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-17"><a href="#cb88-17" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(se_obj, <span class="at">group.by =</span> <span class="st">"annotation_lvl1"</span>, <span class="at">cols =</span> ggsci<span class="sc">::</span><span class="fu">pal_igv</span>()(<span class="dv">50</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="confirm-the-annotation" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="confirm-the-annotation"><span class="header-section-number">3.1</span> Confirm the Annotation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>genes_multiple <span class="ot">&lt;-</span> <span class="fu">str_to_title</span>(<span class="fu">c</span>(<span class="st">"Gfap"</span>, <span class="st">"Aqp4"</span>, <span class="st">"Aldh1l1"</span>,<span class="st">"Slc1a3"</span>, <span class="st">"S100b"</span>, <span class="co"># Astrocytes</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"Flt1"</span>, <span class="st">"Pecam1"</span>, <span class="st">"Cldn5"</span>, <span class="st">"Cdh5"</span>,<span class="st">"Slco1c1"</span>, <span class="co"># Endothilial </span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Krt18"</span>, <span class="st">"Epcam"</span>, <span class="st">"Krt19"</span>, <span class="st">"Krt23"</span>, <span class="co"># Epithilial </span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"Snap25"</span>, <span class="st">"Rbfox3"</span>, <span class="st">"Syt1"</span>, <span class="st">"Grin1"</span>,<span class="st">"Gad1"</span>, <span class="st">"Gad2"</span>, <span class="st">"Slc17a7"</span>, <span class="st">"Map2"</span>, <span class="co"># Neuoron </span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"COL1A1"</span>, <span class="st">"Dcn"</span>, <span class="st">"Col3a1"</span>, <span class="co"># Mesenchymal</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"C1qa"</span>, <span class="st">"Csf1r"</span>, <span class="st">"P2ry12"</span>, <span class="st">"Aif1"</span>, <span class="st">"Tmem119"</span>, <span class="co"># Microglia</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Cd14"</span>, <span class="st">"Psap"</span>, <span class="st">"S100a9"</span>, <span class="st">"Fcer1g"</span>, <span class="co"># Microglia</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Ncr1"</span>, <span class="st">"Klra8"</span>, <span class="st">"Klri2"</span>, <span class="st">"Klrb1c"</span>, <span class="co"># NK</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"Mog"</span>, <span class="st">"Olig2"</span>, <span class="st">"Plp1"</span>, <span class="st">"Cnp"</span>, <span class="co"># Oligodendrocytes</span></span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"Pdgfra"</span>, <span class="st">"Cspg4"</span>, <span class="st">"Sox10"</span>, <span class="st">"Olig1"</span>, <span class="co"># OPcs</span></span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Ptprc"</span>, <span class="st">"Cd3e"</span>, <span class="st">"Cd3d"</span>, <span class="st">"Cd8a"</span>, <span class="st">"Cd3g"</span> <span class="co"># T Cells</span></span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>                                  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="dot-plot" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="dot-plot"><span class="header-section-number">3.1.1</span> Dot plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(<span class="at">object =</span> se_obj, <span class="at">features =</span> genes_multiple, <span class="at">group.by =</span> <span class="st">"annotation_lvl1"</span>) <span class="sc">+</span> </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"viridis"</span>,<span class="at">direction =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>)) <span class="sc">+</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> pct.exp <span class="sc">*</span> <span class="fl">1.1</span>),  <span class="co"># Increase size slightly to create the halo</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> <span class="dv">21</span>,  <span class="co"># Use a circle outline</span></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">fill =</span> <span class="cn">NA</span>,  <span class="co"># Hollow circle</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"black"</span>,  <span class="co"># Outline color (can adjust)</span></span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">stroke =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="clusterd-dot-plot" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="clusterd-dot-plot"><span class="header-section-number">3.1.2</span> Clusterd Dot plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Clustered_DotPlot</span>(<span class="at">seurat_object =</span> se_obj, </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">features =</span> genes_multiple, </span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">cluster_ident =</span> T,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">cluster_feature =</span> T,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">"annotation_lvl1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="3-lung_annotation_files/figure-html/unnamed-chunk-19-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="save-object" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Save Object</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="st">"{path}/4-annotation/objects/se_obj_{model}.qs"</span> <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qsave</span>(se_obj,.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="session-information" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Session Information</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 24.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Madrid
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] glue_1.7.0                  viridis_0.6.5              
 [3] viridisLite_0.4.2           scran_1.30.2               
 [5] reticulate_1.39.0           infercnv_1.18.1            
 [7] qs_0.26.3                   scrubletR_0.1.0            
 [9] ggsci_3.2.0                 SCpubr_2.0.2               
[11] scCustomize_2.1.2           scater_1.30.1              
[13] scuttle_1.12.0              SingleCellExperiment_1.24.0
[15] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[17] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
[19] IRanges_2.36.0              S4Vectors_0.40.2           
[21] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[23] matrixStats_1.4.1           harmony_1.2.1              
[25] Rcpp_1.0.13                 DoubletFinder_2.0.4        
[27] lubridate_1.9.3             forcats_1.0.0              
[29] stringr_1.5.1               dplyr_1.1.4                
[31] purrr_1.0.2                 readr_2.1.5                
[33] tidyr_1.3.1                 tibble_3.2.1               
[35] ggplot2_3.5.1               tidyverse_2.0.0            
[37] Seurat_5.1.0                SeuratObject_5.0.2         
[39] sp_2.1-4                    here_1.0.1                 

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0     bitops_1.0-8             
  [3] httr_1.4.7                RColorBrewer_1.1-3       
  [5] doParallel_1.0.17         tools_4.3.3              
  [7] sctransform_0.4.1         utf8_1.2.4               
  [9] R6_2.5.1                  lazyeval_0.2.2           
 [11] uwot_0.2.2                GetoptLong_1.0.5         
 [13] withr_3.0.1               gridExtra_2.3            
 [15] parallelDist_0.2.6        progressr_0.14.0         
 [17] argparse_2.2.3            cli_3.6.3                
 [19] formatR_1.14              Cairo_1.6-2              
 [21] spatstat.explore_3.3-2    fastDummies_1.7.4        
 [23] sandwich_3.1-1            prismatic_1.1.2          
 [25] labeling_0.4.3            mvtnorm_1.3-1            
 [27] spatstat.data_3.1-2       ggridges_0.5.6           
 [29] pbapply_1.7-2             parallelly_1.38.0        
 [31] limma_3.58.1              rstudioapi_0.16.0        
 [33] generics_0.1.3            shape_1.4.6.1            
 [35] RApiSerialize_0.1.3       gtools_3.9.5             
 [37] ica_1.0-3                 spatstat.random_3.3-2    
 [39] zip_2.3.1                 Matrix_1.6-5             
 [41] futile.logger_1.4.3       ggbeeswarm_0.7.2         
 [43] fansi_1.0.6               abind_1.4-8              
 [45] lifecycle_1.0.4           edgeR_4.0.16             
 [47] multcomp_1.4-26           yaml_2.3.10              
 [49] snakecase_0.11.1          gplots_3.1.3.1           
 [51] SparseArray_1.2.4         Rtsne_0.17               
 [53] paletteer_1.6.0           grid_4.3.3               
 [55] dqrng_0.4.1               promises_1.3.0           
 [57] crayon_1.5.3              miniUI_0.1.1.1           
 [59] lattice_0.22-5            beachmat_2.18.1          
 [61] cowplot_1.1.3             ComplexHeatmap_2.18.0    
 [63] metapod_1.10.1            pillar_1.9.0             
 [65] knitr_1.48                rjson_0.2.23             
 [67] future.apply_1.11.2       codetools_0.2-19         
 [69] leiden_0.4.3.1            spatstat.univar_3.0-1    
 [71] data.table_1.16.0         vctrs_0.6.5              
 [73] png_0.1-8                 spam_2.10-0              
 [75] gtable_0.3.5              rematch2_2.1.2           
 [77] openxlsx_4.2.7.1          xfun_0.47                
 [79] S4Arrays_1.2.1            mime_0.12                
 [81] libcoin_1.0-10            coda_0.19-4.1            
 [83] survival_3.5-8            iterators_1.0.14         
 [85] bluster_1.12.0            statmod_1.5.0            
 [87] TH.data_1.1-2             fitdistrplus_1.2-1       
 [89] ROCR_1.0-11               nlme_3.1-164             
 [91] RcppAnnoy_0.0.22          rprojroot_2.0.4          
 [93] irlba_2.3.5.1             vipor_0.4.7              
 [95] KernSmooth_2.23-22        colorspace_2.1-1         
 [97] ggrastr_1.0.2             tidyselect_1.2.1         
 [99] compiler_4.3.3            BiocNeighbors_1.20.2     
[101] DelayedArray_0.28.0       plotly_4.10.4            
[103] stringfish_0.16.0         caTools_1.18.3           
[105] scales_1.3.0              lmtest_0.9-40            
[107] digest_0.6.37             goftest_1.2-3            
[109] presto_1.0.0              spatstat.utils_3.1-0     
[111] rmarkdown_2.28            XVector_0.42.0           
[113] htmltools_0.5.8.1         pkgconfig_2.0.3          
[115] sparseMatrixStats_1.14.0  fastmap_1.2.0            
[117] rlang_1.1.4               GlobalOptions_0.1.2      
[119] htmlwidgets_1.6.4         shiny_1.9.1              
[121] DelayedMatrixStats_1.24.0 farver_2.1.2             
[123] zoo_1.8-12                jsonlite_1.8.9           
[125] BiocParallel_1.36.0       BiocSingular_1.18.0      
[127] RCurl_1.98-1.16           magrittr_2.0.3           
[129] modeltools_0.2-23         GenomeInfoDbData_1.2.11  
[131] dotCall64_1.1-1           patchwork_1.3.0          
[133] munsell_0.5.1             ape_5.8                  
[135] stringi_1.8.4             zlibbioc_1.48.2          
[137] MASS_7.3-60.0.1           plyr_1.8.9               
[139] parallel_4.3.3            listenv_0.9.1            
[141] ggrepel_0.9.6             deldir_2.0-4             
[143] splines_4.3.3             tensor_1.5               
[145] hms_1.1.3                 circlize_0.4.16          
[147] locfit_1.5-9.10           fastcluster_1.2.6        
[149] igraph_2.0.3              spatstat.geom_3.3-3      
[151] RcppHNSW_0.6.0            reshape2_1.4.4           
[153] ScaledMatrix_1.10.0       futile.options_1.0.1     
[155] evaluate_1.0.0            RcppParallel_5.1.9       
[157] lambda.r_1.2.4            ggprism_1.0.5            
[159] phyclust_0.1-34           tzdb_0.4.0               
[161] foreach_1.5.2             httpuv_1.6.15            
[163] RANN_2.6.2                polyclip_1.10-7          
[165] clue_0.3-65               future_1.34.0            
[167] scattermore_1.2           coin_1.4-3               
[169] rsvd_1.0.5                janitor_2.2.0            
[171] xtable_1.8-4              RSpectra_0.16-2          
[173] later_1.3.2               rjags_4-16               
[175] beeswarm_0.4.0            cluster_2.1.6            
[177] timechange_0.3.0          globals_0.16.3           </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>