---
author: "Mohamed Abdalfttah"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  model: "Default!"
  title: "Neural Circuits - InferCNV `r params$model`  Model"
format:
  html:
    toc: true
    toc-location: left
    number-sections: true
    df-print: paged
editor: 
  render-on-save: true
---

## Script Description

In this script we will annotate the cells from each mouse model separately using two approaches:

1- Canonical Markers

2- Gene expression profile

```         
 "482n1" Lung
 "b16" Melanoma
 "eo7711" Breast
```

## Load Libraries

```{r, eval=FALSE}
#install.packages("devtools")
#devtools::install_github("immunogenomics/presto")
#install.packages('assertthat')
#BiocManager::install("ComplexHeatmap")
#install.packages("quarto")
```

```{r}
library(here)
library(Seurat)
library(tidyverse)
library(DoubletFinder)
library(harmony)
library(scuttle)
library(scater)
library(scCustomize)
library(SCpubr)
library(ggsci)
library(scrubletR)
library(qs)
library(infercnv)
library(reticulate)
library(scran)
library(viridis)
library(glue)
library(quarto)
```

```{r}
path <- "/home/mabdalfttah/projects/neural_circuits/"
```

# Load data

## Read Seurat Object

```{r}
model <- params$model
```

```{r, eval=FALSE}
#model <- "b16"
```

```{r}
se_obj <- "{path}/4-annotation/objects/se_obj_{model}.qs" %>%
  glue::glue() %>%
  here::here() %>%
  qread(.)
se_obj
```

```{r, fig.width=10, fig.height=5}
DimPlot(se_obj, group.by = c("orig.ident", "annotation_lvl1"))
```

## Read gene_order_file

```{r}
gene_order_file <- "{path}/4-annotation/objects/gene_ordering_file.txt" %>%
  glue::glue() %>%
  here::here() %>%
  read.table(., sep = "\t")
rows_gene_order <- gene_order_file$V1
gene_order_file <- gene_order_file[,-1]
rownames(gene_order_file) <- rows_gene_order
```

## Create annotation file

```{r}
annotations_file <- data.frame(row.names = rownames(se_obj@meta.data), celltype = se_obj@meta.data$annotation_lvl1)
annotations_file$celltype <- ifelse(annotations_file$celltype %in% c("Excitatory neurons", "Inhibitory neurons", "Astrocytes"), 
                       "normal", annotations_file$celltype)
unique(annotations_file$celltype)
colnames(annotations_file) <- NULL
```

# Create InferCNV object

```{r}
infercnv_obj = CreateInfercnvObject(raw_counts_matrix=se_obj@assays$RNA$counts,
                                    annotations_file=annotations_file,
                                    gene_order_file=gene_order_file,
                                    ref_group_names=c("normal"))
```

# Run InferCnv

```{r}
output <- "{path}/4-annotation/objects/infercnv_out/{model}" %>%
  glue::glue() %>%
  here::here()
```

```{r}
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                             out_dir=output,  # dir is auto-created for storing outputs
                             cluster_by_groups=T,   # cluster
                             denoise=T,
                             HMM=T
                             )
```

# How it is work? 

InferCNV is a tool used to detect changes in the number of copies of chromosomes, which are common in cancer cells. It doesn't directly identify cancerous cells, but it looks for these chromosome changes that often happen in cancer.

To do this, inferCNV needs a group of normal, healthy cells to use as a reference or baseline. These normal cells should have stable chromosomes, without major changes. In the analysis, you’ll see two main heatmaps:

- Top heatmap: Shows the normal cells, which serve as a comparison. These cells don’t have significant changes, so their heatmap looks pretty stable.

- Bottom heatmap: Shows the cancerous cells. These cells often have a lot of changes in their chromosomes, which inferCNV highlights as red (more copies of certain chromosomes) or blue (fewer copies or losses of certain chromosomes).

```{r}
sessionInfo()
```
